creator-node/src/serviceRegistry.js

```diff
@@ -44,64 +52,197 @@ class ServiceRegistry {
   }
-}
 
-/** Private helper used by ServiceRegistry for initializing libs */
-const initAudiusLibs = async () => {
-  const ethWeb3 = await AudiusLibs.Utils.configureWeb3(
-    config.get('ethProviderUrl'),
-    config.get('ethNetworkId'),
-    /* requiresAccount */ false
-  )
-  const dataWeb3 = await AudiusLibs.Utils.configureWeb3(
-    config.get('dataProviderUrl'),
-    /* chainNetworkId */ null,
-    /* requiresAccount */ false
-  )
-
-  if (!ethWeb3 || !dataWeb3) {
-    throw new Error('Failed to init audiusLibs due to web3 configuration error')
+  /**
+   * Some services require the node server to be running in order to initialize. Run those here.
+   * Specifically:
+   *  - recover node L1 identity (requires node health check from server to return success)
+   *  - initialize SnapbackSM service (requires node L1 identity)
+   *  - register node on L2 URSM contract (requires node L1 identity)
+   */
+  async initServicesThatRequireServer () {
+    if (config.get('isUserMetadataNode')) {
+      return
+    }
+
+    // Cannot progress without recovering spID from node's record on L1 ServiceProviderFactory contract
+    await this._recoverNodeL1Identity()
+
+    // SnapbackSM init (requires L1 identity)
+    await this._initSnapbackSM()
+
+    // L2URSMRegistration (requires L1 identity)
+    await this._registerNodeOnL2URSM()
+
+    this.logInfo(`All services that require server successfully initialized!`)
+  }
+
+  logInfo (msg) {
+    logger.info(`ServiceRegistry || ${msg}`)
+  }
+
+  logError (msg) {
+    logger.error(`ServiceRegistry ERROR || ${msg}`)
+  }
+
+  /**
+   * Poll L1 SPFactory for spID & set spID config once recovered.
+   */
+  async _recoverNodeL1Identity () {
+    const endpoint = config.get('creatorNodeEndpoint')
+
+    const retryTimeoutMs = 5000 // 5sec
+    let attempt = 0
+    while (true) {
+      this.logInfo(`Attempting to recover node L1 identity for ${endpoint} on ${retryTimeoutMs}ms interval || attempt #${++attempt} ...`)
+
+      try {
+        const spID = await this.libs.ethContracts.ServiceProviderFactoryClient.getServiceProviderIdFromEndpoint(
+          endpoint
+        )
+
+        if (spID !== 0) {
+          this.nodeConfig.set('spID', spID)
+          break
+        }
+
+        // Swallow any errors during recovery attempt
+      } catch (e) {
+        this.logError(`RecoverNodeL1Identity Error ${e}`)
+      }
+
+      await utils.timeout(retryTimeoutMs, false)
+    }
+
+    this.logInfo(`Successfully recovered node L1 identity for endpoint ${endpoint} on attempt #${attempt}. spID =  ${this.nodeConfig.get('spID')}`)
+  }
+
+  /**
+   * Wait until URSM contract is deployed, then attempt to register on L2 URSM with infinite retries
+   * Requires L1 identity
+   */
+  async _registerNodeOnL2URSM () {
+    // Wait until URSM contract has been deployed (for backwards-compatibility)
+    let retryTimeoutMs = (this.nodeConfig.get('devMode')) ? 10000 /** 10sec */ : 600000 /* 10min */
+    while (true) {
```

Comment:
can we change these while(true) to like while(!initialized) and set initialized whenever this flow is done? while true just makes me nervous

Reply:
hahaha i had a feeling you would say this - ya sure, changing